executor/src/witgen/jit/witgen_inference.rs

```diff
@@ -386,13 +421,16 @@ impl<'a, T: FieldElement, FixedEval: FixedEvaluator<T>> Evaluator<'a, T, FixedEv
     pub fn evaluate_variable(&self, variable: Variable) -> AffineSymbolicExpression<T, Variable> {
         // If a variable is known and has a compile-time constant value,
         // that value is stored in the range constraints.
-        let rc = self.witgen_inference.range_constraint(variable.clone());
-        if let Some(val) = rc.as_ref().and_then(|rc| rc.try_to_single_value()) {
-            val.into()
-        } else if !self.only_concrete_known && self.witgen_inference.is_known(&variable) {
-            AffineSymbolicExpression::from_known_symbol(variable, rc)
-        } else {
-            AffineSymbolicExpression::from_unknown_variable(variable, rc)
+        let rc = self.witgen_inference.range_constraint(&variable);
+        match (
+            self.witgen_inference.value(&variable),
+            self.only_concrete_known,
+        ) {
+            (Value::Concrete(val), _) => val.into(),
+            (Value::Unknown, _) | (_, true) => {
+                AffineSymbolicExpression::from_unknown_variable(variable, rc)
+            }
+            (Value::Known, false) => AffineSymbolicExpression::from_known_symbol(variable, rc),
         }
```

Comment:
I agree, but what about this one

```suggestion
        let rc = self.witgen_inference.range_constraint(&variable);
        match self.witgen_inference.value(&variable) {
            Value::Concrete(val) => val.into(),
            Value::Unknown => {
                AffineSymbolicExpression::from_unknown_variable(variable, rc)
            }
            Value::Known if self.only_concrete_known => {
                AffineSymbolicExpression::from_unknown_variable(variable, rc)
            }
            Value::Known => AffineSymbolicExpression::from_known_symbol(variable, rc),
        }
```